examples/puppetmaster/test.sh

```diff
@@ -0,0 +1,404 @@
+
+      echo
+      echo "=> Manifest config, HF Token <="
+      converge_node_manifest_hft "$agent_image"
+
+      echo "Tests for '$agent_image': OK"
+    done
+  done
+
+  echo "==="
+  echo "ALL TESTS COMPLETED"
+}
+
+run_in_conjur() {
+  docker-compose exec -T cli "$@"
+}
+
+run_in_puppet() {
+  docker-compose exec -T puppet "$@"
+}
+
+start_services() {
+  docker-compose up -d conjur-https puppet
+}
+
+wait_for_conjur() {
+  docker-compose exec -T conjur conjurctl wait
+}
+
+wait_for_puppetmaster() {
+  echo -n "Waiting on puppetmaster to be ready..."
+  while ! docker-compose exec -T conjur curl -ks https://puppet:8140 >/dev/null; do
+    echo -n "."
+    sleep 2
+  done
+  echo "OK"
+}
+
+get_host_key() {
+  local hostname="$1"
+  run_in_conjur conjur host rotate_api_key -h "$hostname"
+}
+
+get_hf_token() {
+  run_in_conjur conjur hostfactory tokens create --duration-hours 1 inventory | jq -r ".[].token"
+}
+
+install_required_module_dependency() {
+  echo "Installing puppetlabs-registry module dep to server..."
+  docker-compose exec -T puppet puppet module install puppetlabs-registry
+}
+
+setup_conjur() {
+  wait_for_conjur
+  docker-compose exec -T conjur conjurctl account create cucumber || :
+  local api_key=$(docker-compose exec -T conjur conjurctl role retrieve-key cucumber:user:admin | tr -d '\r')
+
+  echo "-----"
+  echo "Starting CLI"
+  echo "-----"
+
+  docker-compose up -d cli
+
+  echo "-----"
+  echo "Logging into the CLI"
+  echo "-----"
+  run_in_conjur conjur authn login -u admin -p "${api_key}"
+
+  echo "-----"
+  echo "Loading Conjur initial policy"
+  echo "-----"
+  run_in_conjur conjur policy load root /src/policy.yml
+  run_in_conjur conjur variable values add inventory/db-password $EXPECTED_PASSWORD  # load the secret's value
+}
+
+revoke_cert_for() {
+  local cert_fqdn="$1"
+  echo "Ensuring clean cert state for $cert_fqdn..."
+
+  # Puppet v5 and v6 CLIs aren't 1:1 compatible so we have to chose the format based
+  # on the server version
+  if [ "${PUPPET_SERVER_TAG:0:1}" == 5 ]; then
+    run_in_puppet puppet cert clean "$cert_fqdn" &>/dev/null || true
+    return
+  fi
+
+  run_in_puppet puppetserver ca revoke --certname "$cert_fqdn" &>/dev/null || true
+  run_in_puppet puppetserver ca clean --certname "$cert_fqdn" &>/dev/null || true
+}
+
+converge_node_agent_apikey() {
+  local agent_image="$1"
+  local node_name="agent-apikey-node"
+  local hostname="${node_name}_$(openssl rand -hex 3)"
+
+  local login="host/$node_name"
+  local api_key=$(get_host_key $node_name)
+  echo "API key for $node_name: $api_key"
+
+  # write the conjurize files to a tempdir so they can be mounted
+  TMPDIR="$PWD/tmp/$(openssl rand -hex 3)"
+  mkdir -p $TMPDIR
+
+  local config_file="$TMPDIR/conjur.conf"
+  local identity_file="$TMPDIR/conjur.identity"
+
+  echo "
+    appliance_url: https://conjur-https:$CONJUR_SERVER_PORT/
+    version: 5
+    account: cucumber
+    cert_file: /etc/ca.crt
+  " > $config_file
```

Comment:
This is probably a don't care, but I think this leaves an extra line with 2 spaces at the bottom of the config file. Why not use heredoc format? Same question for other places where configs are being generated.

Reply:
This is old code but I think the format here makes the logic easier to read and it's still a valid conf file even if it's a bit of an odd spacing within the file